### PR TITLE
CSS 3D transform by matrix3d with translations doesn't take page zoom into account

### PR DESCRIPTION
**Pull request by fujii**
#### fee4b197a6bbb3af225fcd22d2e82af809e1564e
<pre>
CSS 3D transform by matrix3d with translations doesn&apos;t take page zoom into account
<a href="https://bugs.webkit.org/show_bug.cgi?id=164580">https://bugs.webkit.org/show_bug.cgi?id=164580</a>

Reviewed by Antoine Quint.

matrix3d() had a problem if matrix3d() includes translations and page
zoom was applied. The translation component should be treated in the
initial page zoom.
scale3d(zoom, zoom, zoom) × matrix × scale3d(1 / zoom, 1 / zoom, 1 / zoom)

Cherry-picked Blink&apos;s change.
&lt;<a href="https://codereview.chromium.org/2482753002">https://codereview.chromium.org/2482753002</a>&gt;

* LayoutTests/fast/css/getComputedStyle/computed-style-with-zoom-expected.txt:
* LayoutTests/fast/css/getComputedStyle/computed-style-with-zoom.html:
* LayoutTests/transforms/3d/general/cssmatrix-3d-zoom-expected.html: Added.
* LayoutTests/transforms/3d/general/cssmatrix-3d-zoom-expected.png: Removed.
* LayoutTests/transforms/3d/general/cssmatrix-3d-zoom-expected.txt: Removed.
* LayoutTests/transforms/3d/general/cssmatrix-3d-zoom.html: Converted
to a ref-test.
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::matrixTransformValue):
(WebCore::zoomAdjustedNumberValue): Deleted.
* Source/WebCore/css/TransformFunctions.cpp:
(WebCore::transformsForValue):
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::TransformationMatrix::zoom):
* Source/WebCore/platform/graphics/transforms/TransformationMatrix.h:

Canonical link: <a href="https://commits.webkit.org/255416@main">https://commits.webkit.org/255416@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/e5fa37457b1064275b4c84234e546310409b78b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92437 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23030 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102202 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1661 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30043 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84864 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98357 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98099 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1110 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78948 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28044 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/82716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36457 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/16656 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34227 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17828 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3751 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38097 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40001 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36978 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 1]
```diff
@@ -683,9 +678,10 @@ static LayoutRect sizingBox(RenderObject& renderer)
     return box.style().boxSizing() == BoxSizing::BorderBox ? box.borderBoxRect() : box.computedCSSContentBoxRect();
 }
 
-static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& transform, const RenderStyle& style)
+static Ref<CSSFunctionValue> matrixTransformValue(TransformationMatrix transform, const RenderStyle& style)
 {
     RefPtr<CSSFunctionValue> transformValue;
+    transform.zoom(1 / style.effectiveZoom());
```
Comment (fujii) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: This reciprocal is causing imported/w3c/web-platform-tests/css/css-transforms/animation/transform-interpolation-computed-value.html test failure. We should avoid this.

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 2]
```diff
@@ -683,9 +678,10 @@ static LayoutRect sizingBox(RenderObject& renderer)
     return box.style().boxSizing() == BoxSizing::BorderBox ? box.borderBoxRect() : box.computedCSSContentBoxRect();
 }
 
-static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& transform, const RenderStyle& style)
+static Ref<CSSFunctionValue> matrixTransformValue(TransformationMatrix transform, const RenderStyle& style)
 {
     RefPtr<CSSFunctionValue> transformValue;
+    float zoom = style.effectiveZoom();
```
Comment (graouts) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: `auto`

File: Source/WebCore/css/ComputedStyleExtractor.cpp [hunk 3]
```diff
@@ -683,9 +678,10 @@ static LayoutRect sizingBox(RenderObject& renderer)
     return box.style().boxSizing() == BoxSizing::BorderBox ? box.borderBoxRect() : box.computedCSSContentBoxRect();
 }
 
-static Ref<CSSFunctionValue> matrixTransformValue(const TransformationMatrix& transform, const RenderStyle& style)
+static Ref<CSSFunctionValue> matrixTransformValue(TransformationMatrix transform, const RenderStyle& style)
```
Comment (graouts) [inline Source/WebCore/css/ComputedStyleExtractor.cpp]: Why can't we keep the transform `const`? I'm not seeing where we're calling non-const methods on it.

**Review by graouts — CHANGES_REQUESTED** [review]

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 3 above)
Comment (fujii) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: It's a leftover of the previous patch. Will fix. Thank you.

File: Source/WebCore/css/ComputedStyleExtractor.cpp (hunk 2 above)
Comment (fujii) [reply Source/WebCore/css/ComputedStyleExtractor.cpp]: Will fix.

**Review by graouts — APPROVED** [review]

**Comment by fujii** [comment]
EWS reported imported/blink/animations/responsive test failures for Mac and GTK. `waitForCompositor` doesn't work as expected? However, WinCairo passes the tests.
I imported the tests because the original Blink patch modified it. However, WinCairo passes the tests even without this patch.
I'm going to remove the responsive tests.

**Pull request closed without merging**